bug_1292234.patch

```diff
@@ -306,6 +306,111 @@ diff --git a/devtools/server/actors/storage.js b/devtools/server/actors/storage.
+--- /dev/null
++++ b/devtools/server/tests/browser/browser_storage_webext_storage_local.js
+@@ -0,0 +1,89 @@
++/* vim: set ft=javascript ts=2 et sw=2 tw=80: */
++/* Any copyright is dedicated to the Public Domain.
++   http://creativecommons.org/publicdomain/zero/1.0/ */
++
++/* import-globals-from storage-helpers.js */
++
++"use strict";
++
++Services.scriptloader.loadSubScript(
++  "chrome://mochitests/content/browser/devtools/server/tests/browser/storage-helpers.js",
++  this);
++
++const {Toolbox} = require("devtools/client/framework/toolbox");
++
++async function setupExtensionDebuggingToolbox(extensionId, {openToolbox} = {}) {
++  DebuggerServer.init();
++  DebuggerServer.registerAllActors();
++  const transport = DebuggerServer.connectPipe();
++  const client = new DebuggerClient(transport);
++  await client.connect();
++  const front = await client.mainRoot.getAddon({id: extensionId});
++  const target = await front.connect();
++
++  const result = {target, front};
++
++  if (openToolbox) {
++    result.toolbox = await gDevTools.showToolbox(target, null, Toolbox.HostType.WINDOW);
++  }
++
++  return result;
++}
++
++add_task(async function() {
++  async function background() {
++    /* globals browser */
++    browser.test.onMessage.addListener(async (msg, ...args) => {
++      switch (msg) {
++        case "storage-local-set":
++          await browser.storage.local.set(args[0]);
++          break;
++        default:
++          browser.test.fail(`Unexpected test message: ${msg}`);
++      }
++
++      browser.test.sendMessage(`${msg}:done`);
++    });
++
++    browser.test.sendMessage("extension-origin", window.location.origin);
++  }
++  const extension = ExtensionTestUtils.loadExtension({
++    background,
++    manifest: {
++      permissions: ["storage"],
++    },
++    useAddonManager: "temporary",
++  });
++
++  await extension.startup();
++
++  const expectedHost = await extension.awaitMessage("extension-origin");
++
++  const {target} = await setupExtensionDebuggingToolbox(extension.id);
++
++  const storageFront = await target.getFront("storage");
++  const stores = await storageFront.listStores();
++  ok("extensionStorage" in stores, "Should have an extensionStorage store");
++  ok(expectedHost in stores.extensionStorage.hosts,
++     "Should have the expected extension host in the extensionStorage store");
++
++  let data = await stores.extensionStorage.getStoreObjects(expectedHost);
++  Assert.deepEqual(data, {
++    offset: 0, total: 0, data: [],
++    from: stores.extensionStorage.actorID,
++  }, "Got the expected results on empty storage.local");
++
++  extension.sendMessage("storage-local-set", {a: 123});
++  await extension.awaitMessage("storage-local-set:done");
++
++  data = await stores.extensionStorage.getStoreObjects(expectedHost);
++  Assert.deepEqual(data, {
++    // TODO: value is likely a string because of the storage object spec,
++    // but it should actually reflect the actual stored type.
```

Comment:
What value are you referring to here?

Reply:
> The value of the item in the storage local API. Right now we are always treating it like a string, because in the spec that property is a string. We are basically converting it into a string with JSON stringify; technically that is not a string.

> Example: you set a key from the extension that is called mickeyMouse with value `123`. `123` is a number, but the spec says that the value is a string, because we derive the spec from the localStorage Web API whose value is always a string. We convert it as a string to match the spec. If it's a number and we change the number, it should still be a number. If you change `123` and it becomes a string, that is wrong.

> This value is always going to go back and forth in protocol.js level, so will always be stringified there, but from an actor perspective, that property should remain a number. We should ask Mike.

> https://searchfox.org/mozilla-central/source/devtools/shared/specs/storage.js#215 IDB actor uses longString. That would be enough in our case.

> Test case: test extension writes a number in a storage local key and then sends a message to privileged part that updates it with another value, and then be sure in test extension typeof is still a number.

> We don't want to change the data type of the value to a string just from our storage actor for the extension